presto-main/src/test/java/com/facebook/presto/sql/planner/TestDereferencePushDown.java

```diff
@@ -0,0 +1,111 @@
+                                                Optional.of("sum"), functionCall("sum", ImmutableList.of("expr_3")),
+                                                Optional.of("sum7"), functionCall("sum", ImmutableList.of("expr_4")),
+                                                Optional.of("sum8"), functionCall("sum", ImmutableList.of("expr_5")),
+                                                Optional.of("sum9"), functionCall("sum", ImmutableList.of("expr_6"))),
+                                        ImmutableMap.of(),
+                                        Optional.empty(),
+                                        SINGLE,
+                                        anyTree(
+                                            project(ImmutableMap.of("expr_2", expression("field.col1[BIGINT '1']"), "expr_3", expression("field.col0"), "expr_4", expression("field.col1[BIGINT '1'].col1"), "expr_5", expression("CAST(field.col1[BIGINT '2'].col0 AS bigint)"), "expr_6", expression("field.col2.col1")),
+                                                    values("field")))))));
+    }
+
+    @Test
+    public void testDereferencePushdownJoin()
+    {
+        assertPlan("with t1 as ( select * from (values ROW(CAST(ROW(1, 2.0) AS ROW(x BIGINT, y DOUBLE)))) as t (msg) ) select b.msg.x from t1 a, t1 b where a.msg.y = b.msg.y",
+                output(ImmutableList.of("x"),
+                        join(INNER, ImmutableList.of(equiJoinClause("left_y", "right_y")),
+                                anyTree(
+                                        project(ImmutableMap.of("left_y", expression("field.y")),
+                                                values("field"))
+                                ), anyTree(
+                                        project(ImmutableMap.of("right_y", expression("field1.y"), "x", expression("field1.x")),
+                                                values("field1"))))));
+
+        assertPlan("with t1 as ( select * from (values ROW(CAST(ROW(1, 2.0) AS ROW(x BIGINT, y DOUBLE)))) as t (msg) ) select a.msg.y from t1 a join t1 b on a.msg.y = b.msg.y where a.msg.x > bigint '5'",
+                output(ImmutableList.of("left_y"),
+                        join(INNER, ImmutableList.of(equiJoinClause("left_y", "right_y")),
+                                anyTree(
+                                        project(ImmutableMap.of("left_y", expression("field.y")),
+                                                filter("field.x > bigint '5'", values("field")))
+                                ), anyTree(
+                                        project(ImmutableMap.of("right_y", expression("field1.y")),
+                                                values("field1"))))));
+    }
+
+    @Test
+    public void testDereferencePushdownProject()
+    {
+        assertPlan("WITH t1 as ( SELECT * FROM (values ROW(CAST(ROW(1, 2.0) AS ROW(x BIGINT, y DOUBLE)))) as t (msg) ) SELECT msg.x FROM t1 WHERE msg.x > 10",
```

Comment:
Reformat for readability (capitalize SQL keywords, split into multiple lines, replace t1 with t(msg)):

```
        assertPlan("WITH t(msg) AS (SELECT * FROM (VALUES ROW(CAST(ROW(1, 2.0) AS ROW(x BIGINT, y DOUBLE))))) " +
                        "SELECT msg.x FROM t WHERE msg.x > 10",
                output(ImmutableList.of("x"),
                    project(ImmutableMap.of("x", expression("msg.x")),
                            filter("msg.x > BIGINT '10'",
                                    values("msg")))));
```

However, I'd expect msg.x projection to be pushed down, e.g. I'd think the plan would look like

```
Output(x)
   Filter(x > 10)
      Project(x: msg.x)
         Values
``` 

Also, I removed the new rules from the optimizer and re-ran the tests. All tests but testDereferencePushdownJoin succeeded. Could you modify the tests to use queries that are affected by the new rules?

Reply:
let me rework the testcases
predicatePushDown will push down filter just on top of tablescan or values. It is OK we have dereference just on top of filter, as long as ScanFilterAndProject has dereferences, so that virtual nested columns could be generated by leveraging dereferences in connector. What do you think?

Reply:
@zhenxiao That's right. This is a subtle point that's not easy to get from reading the rules code. Perhaps, modify the rules to not do anything for project(filter(tablescan)) and project(filter(values)). It would also help to update the PR description to give examples of queries where dereference pushdown will be successful and show queries where new rules don't make a difference. For the individual rules, consider adding documentation. See com.facebook.presto.sql.planner.iterative.rule.PushProjectionThroughExchange for an example.